### PR TITLE
pod container merge strategic feature.

### LINKED ISSUE #4472: pod template merge patch strategic
## Feature Request

**Is your feature request related to a problem? Please describe:**
I want to add `hostAliases` field to tidbmonitor pod , i find no ways.

**Describe the feature you'd like:**
We can add a pod template merge patch strategic.User can specify pod template like below:
```
spec:
  containers:
  - name: prometheus
    readinessProbe:
      failureThreshold: 500
 hostAliases:
 - hostnames:
    - test.prometheus.com
       ip: 10.96.223.84
```
Operator will merge user custom template and our template , ouput a merge template.

### LINKED ISSUE #4516: Add the Kubernetes' Lifecycle hook support for TiKV and PD
## Feature Request

We want to add the Kubernetes' Lifecycle hook support for TiKV and PD container in the TiDB operator.

**Is your feature request related to a problem? Please describe:**

**Describe the feature you'd like:**

[Kubernetes core/v1.Lifecycle ](https://kubernetes.io/docs/concepts/containers/container-lifecycle-hooks/)  describes actions that the management system should take in response to container lifecycle events. For the PostStart and PreStop lifecycle handlers, management of the container blocks until the action is complete, unless the container process fails, in which case the handler is aborted.

We want to add the Kubernetes' Lifecycle hook support for TiKV and PD container in the TiDB operator.

**Describe alternatives you've considered:**

n/a
**Teachability, Documentation, Adoption, Migration Strategy:**

n/a

### PR DESCRIPTION
**Pull request by mikechengwei**
### What problem does this PR solve?
Close #4472 close #4516.
### What is changed and how does it work?
User can custom pod container configuration. If the container name is not existing, the operator will add this container.If the container  is existing, operator will merge the container configuration into the default .
### Code changes

- [ ] Has Go code change
- [ ] Has CI related scripts change

### Tests

- [ ] Unit test 
- [ ] E2E test 
- [ ] Manual test 
- [ ] No code 

### Side effects

- [ ] Breaking backward compatibility 
- [ ] Other side effects: 

### Related changes

- [ ] Need to cherry-pick to the release branch 
- [ ] Need to update the documentation 

### Release Notes

Please refer to [Release Notes Language Style Guide](https://github.com/pingcap/tidb-operator/blob/master/docs/release-note-guide.md) before writing the release note.

```release-note

```

### CODE REVIEW

File: pkg/apis/pingcap/v1alpha1/types.go [hunk 1]
```diff
@@ -454,6 +454,12 @@ type PDSpec struct {
 	// +optional
 	// +kubebuilder:validation:Enum:="";"v1"
 	StartUpScriptVersion string `json:"startUpScriptVersion,omitempty"`
+	// Containers allows injecting additional containers. This is meant to
+	// allow adding an additional container to an PD pod.
+	// Containers described here modify an operator generated container if they
+	// share the same name and modifications are done via a strategic merge
+	// patch.
+	Containers []corev1.Container `json:"containers,omitempty"`
```
Comment (DanielZhangQD) [inline pkg/apis/pingcap/v1alpha1/types.go:462]: There is already a field `AdditionalContainers` in the `ComponentSpec`, I think we can just reuse it.

Comment (DanielZhangQD) [reply pkg/apis/pingcap/v1alpha1/types.go:462]: BTW, the container name of the components are set by the TiDB Operator, so please help add descriptions in the above field so the user can know how to add new containers or patch existing containers.

Comment (mikechengwei) [reply pkg/apis/pingcap/v1alpha1/types.go:462]: ok

File: pkg/manager/member/pd_member_manager.go [hunk 2]
```diff
@@ -761,7 +763,11 @@ func getNewPDSetForTidbCluster(tc *v1alpha1.TidbCluster, cm *corev1.ConfigMap) (
 	pdContainer.Env = util.AppendEnv(env, basePDSpec.Env())
 	pdContainer.EnvFrom = basePDSpec.EnvFrom()
 	podSpec.Volumes = append(vols, basePDSpec.AdditionalVolumes()...)
-	podSpec.Containers = append([]corev1.Container{pdContainer}, basePDSpec.AdditionalContainers()...)
+	podSpec.Containers, err = MergePatchContainers([]corev1.Container{pdContainer}, basePDSpec.AdditionalContainers())
+	if err != nil {
+		return nil, errors2.Wrap(err, "failed to merge containers spec")
```
Comment (DanielZhangQD) [inline pkg/manager/member/pd_member_manager.go:768]: Why introduce the errors2 here? Can we just use the same error-generating code as the other ones in the same file to keep the same style?

File: pkg/manager/member/pd_member_manager_test.go [hunk 3]
```diff
@@ -183,6 +183,23 @@ func TestPDMemberManagerSyncCreate(t *testing.T) {
 			pdPeerSvcCreated: false,
 			setCreated:       false,
 		},
+		{
+			name: "patch pod container",
```
Comment (DanielZhangQD) [inline pkg/manager/member/pd_member_manager_test.go:187]: Please cover the case where the fields in the additional container override the fields generated by TiDB Operator.

Comment (DanielZhangQD) [reply pkg/manager/member/pd_member_manager_test.go:187]: Can we check the final result of the PD container?

Comment (DanielZhangQD) [reply pkg/manager/member/pd_member_manager_test.go:187]: And also cover the case with a new additional container.

File: pkg/manager/member/utils.go [hunk 4]
```diff
@@ -479,3 +482,59 @@ func TiKVStoreIDFromStatus(tc *v1alpha1.TidbCluster, podName string) (uint64, er
 	}
 	return 0, ErrNotFoundStoreID
 }
+
+// MergePatchContainers adds patches to base using a strategic merge patch and
+// iterating by container name, failing on the first error
+func MergePatchContainers(base, patches []corev1.Container) ([]corev1.Container, error) {
```
Comment (DanielZhangQD) [inline pkg/manager/member/utils.go:488]: Please add UT for this function.

File: pkg/manager/member/tikv_member_manager.go [hunk 5]
```diff
@@ -613,7 +615,12 @@ func getNewTiKVSetForTidbCluster(tc *v1alpha1.TidbCluster, cm *corev1.ConfigMap)
 	podSpec.Volumes = append(vols, baseTiKVSpec.AdditionalVolumes()...)
 	podSpec.SecurityContext = podSecurityContext
 	podSpec.InitContainers = append(initContainers, baseTiKVSpec.InitContainers()...)
-	podSpec.Containers = append(containers, baseTiKVSpec.AdditionalContainers()...)
+
+	podSpec.Containers, err = MergePatchContainers(containers, baseTiKVSpec.AdditionalContainers())
```
Comment (DanielZhangQD) [inline pkg/manager/member/tikv_member_manager.go:619]: Please also update the code for TiDB Monitor, tidbngmonitoring, DM master, DM worker, Discovery, TiCDC, and Pump.

Comment (mikechengwei) [reply pkg/manager/member/tikv_member_manager.go:620]: ok

File: pkg/apis/pingcap/v1alpha1/types.go [hunk 6]
```diff
@@ -981,6 +981,11 @@ type ComponentSpec struct {
 	InitContainers []corev1.Container `json:"initContainers,omitempty"`
 
 	// Additional containers of the component.
+	// Containers described here modify an operator generated container if they
+	// share the same name and modifications are done via a strategic merge
+	// patch.
+	// If additional container name is not contained in operator generated containers, it will add into
+	// pod containers directly.
```
Comment (DanielZhangQD) [inline pkg/apis/pingcap/v1alpha1/types.go:988]:
```suggestion
	// If the container names in this field match with the ones generated by
	// TiDB Operator, the container configurations will be merged into the
	// containers generated by TiDB Operator via strategic merge patch.
	// If the container names in this field do not  match with the ones 
	// generated by TiDB Operator, the container configurations will be
	// appended to the Pod container spec directly.
```

File: pkg/manager/member/pd_member_manager_test.go (hunk 3 above)
Comment (mikechengwei) [reply pkg/manager/member/pd_member_manager_test.go:187]: `TestPDMemberManagerSyncCreate` can't check final result, i will add  code into `TestPDMemberManagerSyncUpdate`.

File: pkg/manager/member/ticdc_member_manager.go [hunk 7]
```diff
@@ -489,7 +489,13 @@ done
 	}
 
 	podSpec := baseTiCDCSpec.BuildPodSpec()
-	podSpec.Containers = []corev1.Container{ticdcContainer}
+
+	var err error
+	podSpec.Containers, err = MergePatchContainers([]corev1.Container{ticdcContainer}, baseTiCDCSpec.AdditionalContainers())
+	if err != nil {
+		return nil, fmt.Errorf("ticdc[%s/%s] failed to merge containers spec , error: %v", ns, stsName, err)
```
Comment (DanielZhangQD) [inline pkg/manager/member/ticdc_member_manager.go:496]:
```suggestion
		return nil, fmt.Errorf("failed to merge containers spec for TiCDC of [%s/%s], error: %v", ns, tcName, err)
```

File: pkg/manager/member/tidb_discovery_manager.go [hunk 8]
```diff
@@ -230,7 +230,12 @@ func (m *realTidbDiscoveryManager) getTidbDiscoveryDeployment(obj metav1.Object)
 			},
 		},
 	})
-	podSpec.Containers = append(podSpec.Containers, baseSpec.AdditionalContainers()...)
+
+	var err error
+	podSpec.Containers, err = MergePatchContainers(podSpec.Containers, baseSpec.AdditionalContainers())
+	if err != nil {
+		return nil, fmt.Errorf("discovery[%s/%s] failed to merge containers spec , error: %v", meta.Namespace, meta.Name, err)
```
Comment (DanielZhangQD) [inline pkg/manager/member/tidb_discovery_manager.go:237]:
```suggestion
		return nil, fmt.Errorf("failed to merge containers spec for Discovery of [%s/%s], error: %v", meta.Namespace, meta.Name, err)
```

File: pkg/manager/member/tidb_member_manager.go [hunk 9]
```diff
@@ -895,7 +897,13 @@ func getNewTiDBSetForTidbCluster(tc *v1alpha1.TidbCluster, cm *corev1.ConfigMap)
 	containers = append(containers, c)
 
 	podSpec := baseTiDBSpec.BuildPodSpec()
-	podSpec.Containers = append(containers, baseTiDBSpec.AdditionalContainers()...)
+
+	var err error
+	podSpec.Containers, err = MergePatchContainers(containers, baseTiDBSpec.AdditionalContainers())
+	if err != nil {
+		return nil, errors2.Wrap(err, "failed to merge containers spec")
```
Comment (DanielZhangQD) [inline pkg/manager/member/tidb_member_manager.go:904]: Please remove errors2 and optimize the error message according to the other comments.

File: pkg/manager/member/tikv_member_manager.go [hunk 10]
```diff
@@ -614,7 +616,12 @@ func getNewTiKVSetForTidbCluster(tc *v1alpha1.TidbCluster, cm *corev1.ConfigMap)
...
+	if err != nil {
+		return nil, errors2.Wrap(err, "failed to merge containers spec")
```
Comment (DanielZhangQD) [inline pkg/manager/member/tikv_member_manager.go:622]: Please remove errors2 and optimize the error message according to the other comments.

File: pkg/manager/member/tiflash_member_manager.go [hunk 11]
```diff
@@ -558,7 +558,12 @@ sed -i s/PD_ADDR/${result}/g /data0/proxy.toml
 		return nil, err
 	}
 	podSpec.Containers = append([]corev1.Container{tiflashContainer}, containers...)
-	podSpec.Containers = append(podSpec.Containers, baseTiFlashSpec.AdditionalContainers()...)
+
+	podSpec.Containers, err = MergePatchContainers(podSpec.Containers, baseTiFlashSpec.AdditionalContainers())
+	if err != nil {
+		return nil, fmt.Errorf("tiflash[%s:%s] failed to merge containers spec,err:%v", ns, setName, err)
```
Comment (DanielZhangQD) [inline pkg/manager/member/tiflash_member_manager.go:564]:
```suggestion
		return nil, fmt.Errorf("failed to merge containers spec for TiFlash of [%s:%s], err: %v", ns, tcName, err)
```

File: pkg/manager/member/utils.go [hunk 12]
```diff
@@ -21,6 +21,9 @@ import (
 	"strings"
 	"time"
 
+	errors2 "github.com/pingcap/errors"
```
Comment (DanielZhangQD) [inline pkg/manager/member/utils.go:24]: Please remove errors2.

File: pkg/manager/tidbngmonitoring/ng_monitoring_manager.go [hunk 13]
```diff
@@ -390,7 +391,11 @@ func GenerateNGMonitoringStatefulSet(tngm *v1alpha1.TidbNGMonitoring, tc *v1alph
 	builder.PodTemplateSpecBuilder().ContainerBuilder(nmContainerName).AddVolumeMounts(spec.AdditionalVolumeMounts()...)
 	builder.PodTemplateSpecBuilder().AddVolumes(spec.AdditionalVolumes()...)
 	// additional containers
-	builder.PodTemplateSpecBuilder().AddContainers(spec.AdditionalContainers()...)
+	builder.PodTemplateSpecBuilder().Get().Spec.Containers, err = member.MergePatchContainers(builder.PodTemplateSpecBuilder().Get().Spec.Containers, spec.AdditionalContainers())
+	if err != nil {
+		return nil, fmt.Errorf("tngm[%s/%s] failed to merge containers spec , error: %v", tngm.Namespace, tngm.Name, err)
```
Comment (DanielZhangQD) [inline pkg/manager/tidbngmonitoring/ng_monitoring_manager.go:396]: Please optimize the error message.

File: pkg/monitor/monitor/util.go [hunk 14]
```diff
@@ -1235,7 +1236,11 @@ func getMonitorStatefulSet(sa *core.ServiceAccount, secret *core.Secret, monitor
 	}
 	additionalContainers := monitor.Spec.AdditionalContainers
 	if len(additionalContainers) > 0 {
-		statefulSet.Spec.Template.Spec.Containers = append(statefulSet.Spec.Template.Spec.Containers, additionalContainers...)
+		var err error
+		statefulSet.Spec.Template.Spec.Containers, err = member.MergePatchContainers(statefulSet.Spec.Template.Spec.Containers, additionalContainers)
+		if err != nil {
+			return nil, fmt.Errorf("tidbmonitor[%s/%s] failed to merge containers spec , error: %v", monitor.Namespace, monitor.Name, err)
```
Comment (DanielZhangQD) [inline pkg/monitor/monitor/util.go:1242]: ditto

File: examples/podpatch/tidb-monitor.yaml [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+    baseImage: pingcap/tidb-monitor-initializer
+    version: v5.4.1
+  reloader:
+    baseImage: pingcap/tidb-monitor-reloader
+    version: v1.0.1
+  prometheusReloader:
+    baseImage: quay.io/prometheus-operator/prometheus-config-reloader
+    version: v0.49.0
+  imagePullPolicy: IfNotPresent
+  additionalContainers:
```
Comment (DanielZhangQD) [inline examples/podpatch/tidb-monitor.yaml:25]: Can the monitor Pod be running with this config?

File: pkg/manager/tidbngmonitoring/ng_monitoring_manager.go [hunk 16]
```diff
@@ -390,7 +391,11 @@ func GenerateNGMonitoringStatefulSet(tngm *v1alpha1.TidbNGMonitoring, tc *v1alph
 	builder.PodTemplateSpecBuilder().ContainerBuilder(nmContainerName).AddVolumeMounts(spec.AdditionalVolumeMounts()...)
 	builder.PodTemplateSpecBuilder().AddVolumes(spec.AdditionalVolumes()...)
 	// additional containers
-	builder.PodTemplateSpecBuilder().AddContainers(spec.AdditionalContainers()...)
+	builder.PodTemplateSpecBuilder().Get().Spec.Containers, err = member.MergePatchContainers(builder.PodTemplateSpecBuilder().Get().Spec.Containers, spec.AdditionalContainers())
+	if err != nil {
+		return nil, fmt.Errorf("failed to merge containers spec for TNGM of [%s/%s] , error: %v", tngm.Namespace, tngm.Name, err)
```
Comment (DanielZhangQD) [inline pkg/manager/tidbngmonitoring/ng_monitoring_manager.go:396]:
```suggestion
		return nil, fmt.Errorf("failed to merge containers spec for TNGM of [%s/%s], error: %v", tngm.Namespace, tngm.Name, err)
```

File: pkg/manager/member/utils.go [hunk 17]
```diff
@@ -479,3 +480,59 @@ func TiKVStoreIDFromStatus(tc *v1alpha1.TidbCluster, podName string) (uint64, er
...
+		patchContainer, ok := containersByName[container.Name]
+		if !ok {
+			// This container didn't need to be patched.
+			out = append(out, container)
+			continue
+		}
+
+		containerBytes, err := json.Marshal(container)
+		if err != nil {
+			return nil, fmt.Errorf("failed to marshal JSON for container %s,error:%v", container.Name, err)
```
Comment (DanielZhangQD) [inline pkg/manager/member/utils.go:505]:
```suggestion
			return nil, fmt.Errorf("failed to marshal JSON for container %s, error: %v", container.Name, err)
```

File: pkg/manager/member/utils.go [hunk 18]
```diff
@@ -479,3 +480,59 @@ func TiKVStoreIDFromStatus(tc *v1alpha1.TidbCluster, podName string) (uint64, er
...
+		}
+
+		patchBytes, err := json.Marshal(patchContainer)
+		if err != nil {
+			return nil, fmt.Errorf("failed to marshal JSON for patch container %s,error:%v", container.Name, err)
```
Comment (DanielZhangQD) [inline pkg/manager/member/utils.go:510]:
```suggestion
			return nil, fmt.Errorf("failed to marshal JSON for patch container %s, error: %v", container.Name, err)
```

File: pkg/manager/member/utils.go [hunk 19]
```diff
@@ -479,3 +480,59 @@ func TiKVStoreIDFromStatus(tc *v1alpha1.TidbCluster, podName string) (uint64, er
...
+		}
+
+		// Calculate the patch result.
+		jsonResult, err := strategicpatch.StrategicMergePatch(containerBytes, patchBytes, corev1.Container{})
+		if err != nil {
+			return nil, fmt.Errorf("failed to generate merge patch for container %s,error:%v", container.Name, err)
```
Comment (DanielZhangQD) [inline pkg/manager/member/utils.go:516]:
```suggestion
			return nil, fmt.Errorf("failed to generate merge patch for container %s, error: %v", container.Name, err)
```

File: pkg/manager/member/utils.go [hunk 20]
```diff
@@ -479,3 +480,59 @@ func TiKVStoreIDFromStatus(tc *v1alpha1.TidbCluster, podName string) (uint64, er
...
+		}
+
+		var patchResult corev1.Container
+		if err := json.Unmarshal(jsonResult, &patchResult); err != nil {
+			return nil, fmt.Errorf("failed to unmarshal merged container %s,error:%v", container.Name, err)
```
Comment (DanielZhangQD) [inline pkg/manager/member/utils.go:521]:
```suggestion
			return nil, fmt.Errorf("failed to unmarshal merged container %s, error: %v", container.Name, err)
```

File: pkg/manager/member/tidb_member_manager.go [hunk 21]
```diff
@@ -895,7 +895,13 @@ func getNewTiDBSetForTidbCluster(tc *v1alpha1.TidbCluster, cm *corev1.ConfigMap)
 	containers = append(containers, c)
 
 	podSpec := baseTiDBSpec.BuildPodSpec()
-	podSpec.Containers = append(containers, baseTiDBSpec.AdditionalContainers()...)
+
+	var err error
+	podSpec.Containers, err = MergePatchContainers(containers, baseTiDBSpec.AdditionalContainers())
+	if err != nil {
+		return nil, fmt.Errorf("failed to merge containers spec for TiDB of [%s:%s],error: %v", ns, tcName, err)
```
Comment (DanielZhangQD) [inline pkg/manager/member/tidb_member_manager.go:902]:
```suggestion
		return nil, fmt.Errorf("failed to merge containers spec for TiDB of [%s:%s], error: %v", ns, tcName, err)
```

File: pkg/manager/member/tikv_member_manager.go [hunk 22]
```diff
@@ -614,7 +614,12 @@ func getNewTiKVSetForTidbCluster(tc *v1alpha1.TidbCluster, cm *corev1.ConfigMap)
 	podSpec.Volumes = append(vols, baseTiKVSpec.AdditionalVolumes()...)
 	podSpec.SecurityContext = podSecurityContext
 	podSpec.InitContainers = append(initContainers, baseTiKVSpec.InitContainers()...)
-	podSpec.Containers = append(containers, baseTiKVSpec.AdditionalContainers()...)
+
+	podSpec.Containers, err = MergePatchContainers(containers, baseTiKVSpec.AdditionalContainers())
+	if err != nil {
+		return nil, fmt.Errorf("failed to merge containers spec for TiKV of [%s:%s],error: %v", ns, tcName, err)
```
Comment (DanielZhangQD) [inline pkg/manager/member/tikv_member_manager.go:620]:
```suggestion
		return nil, fmt.Errorf("failed to merge containers spec for TiKV of [%s:%s], error: %v", ns, tcName, err)
```

**Review by DanielZhangQD — APPROVED** [review]
LGTM

File: pkg/manager/member/tidb_member_manager.go [hunk 23]
```diff
@@ -895,7 +895,13 @@ func getNewTiDBSetForTidbCluster(tc *v1alpha1.TidbCluster, cm *corev1.ConfigMap)
 	containers = append(containers, c)
 
 	podSpec := baseTiDBSpec.BuildPodSpec()
-	podSpec.Containers = append(containers, baseTiDBSpec.AdditionalContainers()...)
+
+	var err error
+	podSpec.Containers, err = MergePatchContainers(containers, baseTiDBSpec.AdditionalContainers())
+	if err != nil {
+		return nil, fmt.Errorf("failed to merge containers spec for TiDB of [%s:%s], error: %v", ns, tcName, err)
```
Comment (csuzhangxc) [inline pkg/manager/member/tidb_member_manager.go:902]: `%s/%s` not `%s:%s` as others did?

File: pkg/manager/member/tiflash_member_manager.go [hunk 24]
```diff
@@ -558,7 +558,12 @@ sed -i s/PD_ADDR/${result}/g /data0/proxy.toml
 		return nil, err
 	}
 	podSpec.Containers = append([]corev1.Container{tiflashContainer}, containers...)
-	podSpec.Containers = append(podSpec.Containers, baseTiFlashSpec.AdditionalContainers()...)
+
+	podSpec.Containers, err = MergePatchContainers(podSpec.Containers, baseTiFlashSpec.AdditionalContainers())
+	if err != nil {
+		return nil, fmt.Errorf("failed to merge containers spec for TiFlash of [%s:%s], err: %v", ns, tcName, err)
```
Comment (csuzhangxc) [inline pkg/manager/member/tiflash_member_manager.go:564]: ditto?

File: pkg/manager/member/tikv_member_manager.go [hunk 25]
```diff
@@ -614,7 +614,12 @@ func getNewTiKVSetForTidbCluster(tc *v1alpha1.TidbCluster, cm *corev1.ConfigMap)
 	podSpec.Volumes = append(vols, baseTiKVSpec.AdditionalVolumes()...)
 	podSpec.SecurityContext = podSecurityContext
 	podSpec.InitContainers = append(initContainers, baseTiKVSpec.InitContainers()...)
-	podSpec.Containers = append(containers, baseTiKVSpec.AdditionalContainers()...)
+
+	podSpec.Containers, err = MergePatchContainers(containers, baseTiKVSpec.AdditionalContainers())
+	if err != nil {
+		return nil, fmt.Errorf("failed to merge containers spec for TiKV of [%s:%s], error: %v", ns, tcName, err)
```
Comment (csuzhangxc) [inline pkg/manager/member/tikv_member_manager.go:620]: ditto?

File: pkg/manager/member/utils_test.go [hunk 26]
```diff
@@ -839,3 +838,143 @@ func TestTiKVLessThanV50(t *testing.T) {
...
+			name:    "port name conflict",
+			base:    []v1.Container{build("c1", "image:A", port1A)},
+			patches: []v1.Container{build("c1", "image:A", port2A)},
+			result:  []v1.Container{build("c1", "image:A", port2A, port1A)}, // port ordering doesn't matter here
+		},
+		{
+			name:    "port value conflict",
+			base:    []v1.Container{build("c1", "image:A", port1A)},
+			patches: []v1.Container{build("c1", "image:A", port1B)},
+			result:  []v1.Container{build("c1", "image:A", port1B)}, // port ordering doesn't matter here
```
Comment (csuzhangxc) [inline pkg/manager/member/utils_test.go:921]: useless comment `// port ordering doesn't matter here`?

**Review by csuzhangxc — APPROVED** [review]

**Pull request merged**